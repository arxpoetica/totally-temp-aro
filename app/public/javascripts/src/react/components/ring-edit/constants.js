export default Object.freeze({
  START_STATE: 'START_STATE',
  STARTED: 'STARTED',
  COMPLETED: 'COMPLETED'
})
