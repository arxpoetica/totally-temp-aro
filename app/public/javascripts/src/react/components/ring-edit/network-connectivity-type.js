export default Object.freeze({
  none: { id: 'none', displayName: 'None' },
  snapToEdge: { id: 'snapToEdge', displayName: 'Snap to Edge' },
  snapToWormhole: { id: 'snapToWormhole', displayName: 'Snap to Location Connector' }
})
