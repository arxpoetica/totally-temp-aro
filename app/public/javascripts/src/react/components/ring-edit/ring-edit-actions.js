/* globals */
import Actions from '../../common/actions'
import AroHttp from '../../common/aro-http'
import RingUtils from './ring-utils'
import Ring from '../../common/ring'

function setSelectedRingId (ringId) {
  return {
    type: Actions.RING_SET_SELECTED_RING_ID,
    payload: ringId
  }
}

function newRing (planId, userId) {
  return (dispatch) => {
    /*
    var promisses = []
    rings.forEach(ring => {
      promisses.push(AroHttp.post(`/service/plan/${planId}/ring-config`, ring.getDataExport()))
    })
    */
    // Promise.all(promisses)
    AroHttp.post(`/service/plan/${planId}/ring-config`, {})
      .then(result => {
      // ToDo protect against fail returns
        var ring = new Ring(result.data.id)
        dispatch({
          type: Actions.RING_ADD_RINGS,
          payload: [ring]
        })
      }).catch(err => console.error(err))
  }
}

function removeAllRings () {
  return {
    type: Actions.RING_REMOVE_ALL_RINGS
  }
}

function addNode (ring, feature, planId, userId) {
  return (dispatch) => {
    var ringClone = ring.clone()
    var featureIndex = ringClone.nodes.findIndex((ele) => ele.objectId == feature.objectId)

    if (featureIndex != -1) return
    ringClone.addNode(feature)

    // todo make ring update action
    AroHttp.put(`/service/plan/${planId}/ring-config/${ring.id}`, ringClone.getDataExport())
      .then(result => {
      // ToDo protect against fail returns
        dispatch({
          type: Actions.RING_UPDATE_RING,
          payload: ringClone
        })
      }).catch(err => console.error(err))
  }
}

function removeNode (ring, featureId, planId, userId) {
  return (dispatch) => {
    var ringClone = ring.clone()
    var featureIndex = ringClone.nodes.findIndex((ele) => ele.objectId == featureId)
    if (featureIndex == -1) return
    ringClone.removeNode(featureId)

    // todo make ring update action
    AroHttp.put(`/service/plan/${planId}/ring-config/${ring.id}`, ringClone.getDataExport())
      .then(result => {
      // ToDo protect against fail returns
        dispatch({
          type: Actions.RING_UPDATE_RING,
          payload: ringClone
        })
      }).catch(err => console.error(err))
  }
}

function removeRing (ringId, planId, userId) {
  return (dispatch) => {
    AroHttp.delete(`/service/plan/${planId}/ring-config/${ringId}`)
      .then(result => {
      // ToDo protect against fail returns
        dispatch({
          type: Actions.RING_REMOVE_RING,
          payload: ringId
        })
      }).catch(err => console.error(err))
  }
}

function saveRingChangesToServer (ring, planId, userId) {
  return (dispatch) => {
    AroHttp.put(`/service/plan/${planId}/ring-config/${ring.id}`, ring.getDataExport())
    .then(result => {
      // ToDo protect against fail returns
      /* // no need to update state, we don't need a redraw
      dispatch({
        type: Actions.RING_UPDATE_RING,
        payload: ringClone
      })
      */
    }).catch(err => console.error(err))
  }
}

function onFeatureSelected (features) {
  // this may be a bit funky, revisit this

  return (dispatch, getState) => {
    const state = getState()

    if (state.ringEdit.selectedRingId &&
      state.ringEdit.rings.hasOwnProperty(state.ringEdit.selectedRingId) &&
      features.equipmentFeatures &&
      features.equipmentFeatures.length > 0
    ) {
      // add selected feature to selected ring
      // OR delete selected feature from selected ring
      var validNodes = features.equipmentFeatures.filter(feature => feature._data_type.includes('central_office'))
      if (validNodes.length == 0) return // {type:null}
      var feature = { ...validNodes[0] }
      feature.objectId = feature.object_id
      var ring = state.ringEdit.rings[state.ringEdit.selectedRingId]
      var featureIndex = ring.nodes.findIndex((ele) => ele.objectId == feature.objectId)
      const planId = state.plan.activePlan.id
      const userId = state.user.loggedInUser.id

      if (featureIndex != -1) {
        // remove node
        dispatch(removeNode(ring, feature.objectId, planId, userId))
      } else {
        // add node
        // get feature lat long

        RingUtils.getEquipmentDataPromise(feature.objectId, planId, userId)
          .then(result => {
            feature.data = result.data
            dispatch(addNode(ring, feature, planId, userId))
          }).catch(err => console.error(err))
      }
    }
  }
}

function loadRings (planId) {
  return (dispatch, getState) => {
    const state = getState()
    const userId = state.user.loggedInUser.id
    AroHttp.get(`/service/plan/${planId}/ring-config?planId=${planId}`)
      .then(result => {
        dispatch({
          type: Actions.RING_REMOVE_ALL_RINGS
        })
        var rings = []
        result.data.forEach(ringData => {
          rings.push(Ring.parseData(ringData, planId, userId))
        })
        dispatch({
          type: Actions.RING_ADD_RINGS,
          payload: rings
        })
      })
      .catch(err => console.error(err))
  }
}

export default {
  setSelectedRingId,
  newRing,
  removeRing,
  removeNode,
  removeAllRings,
  onFeatureSelected,
  loadRings,
  addNode,
  removeNode, 
  saveRingChangesToServer
}
