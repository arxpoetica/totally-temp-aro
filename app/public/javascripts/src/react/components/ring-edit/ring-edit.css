.ring-table-item-selected, table td.ring-table-item-selected{
  padding-left: 0px;
  padding-right: 0px;
  padding-bottom: 0px;
}

.ring-sub-table{
  padding-left: 21px;
  padding-bottom: 16px;
  background-color: #ffffff;
  border-left: 1px #eeeeee dashed;
}

.ring-sub-table > table{
  border-left: 1px #eeeeee solid;
  margin-bottom: 0px;
}

.ring-table-item-title{
  
}

.ring-table-item-selected .ring-table-item-title{
  padding: .3rem .3rem 0px .3rem;
}

.ring-table-item-title-selected{
  border-bottom: 2px solid #888888;
  font-weight: bold;
}

.ring-del-btn{
  float: right;
  margin-bottom: 0.3rem;
}

.ring-text-inp{
  width: 50%;
  margin-bottom: 0.3rem;
}