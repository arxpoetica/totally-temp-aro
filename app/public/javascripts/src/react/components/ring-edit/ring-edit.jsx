/* global google, swal */
import React, { Component } from 'react'
import reduxStore from '../../../redux-store'
import wrapComponentWithProvider from '../../common/provider-wrapped-component'
import { PropTypes } from 'prop-types'
import { createSelector } from 'reselect'
import ringActions from './ring-edit-actions.js'
import './ring-edit.css'
import RingStatusTypes from './constants'
import Ring from '../../common/ring'
import RingOptionsBasic from './ring-options-basic.jsx'
import RingOptionsConnectivityDefinition from './ring-options-connectivity-definition.jsx'

export class RingEdit extends Component {
  constructor (props) {
    super(props)

    this.createdMapObjects = []
    this.mapObjectListeners = []
    this.canEdit = true
  }

  render () {
    this.canEdit = (this.props.status === RingStatusTypes.START_STATE)
    this.drawRings()
    return <div>
      {this.renderAddButton()}
      <div className='m-2 p-2'>
        <h4>Rings</h4>
        <table className='table table-sm table-striped'>
          <tbody>
            {this.renderRingRows(this.props.rings)}
          </tbody>
        </table>
      </div>
      <RingOptionsBasic initialValues={this.props.ringOptionsBasic} enableReinitialize />
      <RingOptionsConnectivityDefinition enableReinitialize />
    </div>
  }

  renderAddButton () {
    if (this.canEdit) {
      return <div className='m-2 p-2'>
        <button id='btnRingNewRing'
          className='btn btn-sm btn-light'
          onClick={() => this.requestAddNewRing()}>
          <i className='fas fa-pencil-alt' /> Add Ring
        </button>
      </div>
    } else {
      return ''
    }
  }

  renderRingRows (rings) {
    // reverse order, new ones first
    var revOrder = []
    Object.keys(rings).sort().map((key) => (
      revOrder.unshift(rings[key])
    ))
    var jsx = []
    revOrder.forEach((ring) => {
      jsx.push(this.renderRingRow(ring))
    })
    return jsx
  }

  renderRingRow (ring) {
    if (ring.id === this.props.selectedRingId) {
      // selected ring
      return <tr key={ring.id}>
        <td className='ring-table-item-selected'>
          <div className='ring-table-item-title ring-table-item-title-selected clearfix'>
            {ring.name}

            {this.canEdit
              ? <button
                id={`btnRingDel_${ring.id}`}
                className='btn btn-sm btn-outline-danger ring-del-btn'
                onClick={() => this.requestDeleteRing(ring)}
                data-toggle='tooltip' data-placement='bottom' title='Delete'>
                <i className='fa ei-button-icon ng-scope fa-trash-alt' />
              </button>
              : null
            }

            {this.canEdit
              ? <input
                id={`inpRingName_${ring.id}`}
                type='text'
                className='form-control form-control-sm ring-text-inp'
                placeholder='rename'
                onBlur={event => this.renameRing(ring.id, event.target.value)}
                onKeyDown={event => {
                  if (event.key === 'Enter') {
                    this.renameRing(ring.id, event.target.value)
                  }
                }}
              />
              : null
            }

          </div>
          <div className='ring-sub-table'>
            <table className='table table-sm table-striped'>
              <tbody>
                {
                  ring.nodes.map((node, index) => (
                    <tr className='m-2 p-2' key={ring.id + '_' + node.objectId}>
                      <td>
                        {node.siteClli || node.objectId}

                        {this.canEdit
                          ? <button id={`btnNodeDel_${ring.id}-${node.objectId}`} className='btn btn-sm btn-outline-danger ring-del-btn'
                            onClick={() => this.deleteNode(ring, node.objectId)}
                            data-toggle='tooltip' data-placement='bottom' title='Delete'>
                            <i className='fa ei-button-icon ng-scope fa-trash-alt' />
                          </button>
                          : null
                        }

                      </td>
                    </tr>
                  ))
                }
              </tbody>
            </table>
          </div>
        </td>
      </tr>
    } else {
      return <tr key={ring.id}>
        <td>
          <div id={`btnRingSelect_${ring.id}`} onClick={() => this.props.setSelectedRingId(ring.id)}>
            {ring.name}
          </div>
        </td>
      </tr>
    }
  }

  requestAddNewRing () {
    const planId = this.props.plan.activePlan.id
    const userId = this.props.user.loggedInUser.id
    this.props.newRing(planId, userId)
  }

  renameRing (ringId, val) {
    const planId = this.props.plan.activePlan.id
    const userId = this.props.user.loggedInUser.id
    this.props.renameRing(this.props.rings[ringId], val, planId, userId)
  }

  drawRings () {
    // clear prev lines
    this.clearRendering()

    // for (let [ringId, ring] of Object.entries(this.props.rings)) {
    Object.keys(this.props.rings).forEach(ringId => {
      const ring = this.props.rings[ringId]
      ringId = parseInt(ringId)
      if (ring.nodes.length > 0) {
        var pathCoords = []

        ring.nodes.forEach(node => {
          const coords = node.data.geometry.coordinates
          pathCoords.push({ lat: coords[1], lng: coords[0] })
        })

        var polygonOptions = {
          strokeColor: '#888888',
          strokeOpacity: 0.8,
          strokeWeight: 1,
          fillColor: '#888888',
          fillOpacity: 0.3,
          editable: false
        }

        if (ringId === this.props.selectedRingId) {
          polygonOptions = {
            strokeColor: '#FF1493',
            strokeOpacity: 0.8,
            strokeWeight: 2,
            fillColor: '#FF1493',
            fillOpacity: 0.4,
            editable: this.canEdit
          }
        }

        ring.linkData.forEach(link => {
          var polygon = new google.maps.Polygon({
            paths: link.geom,
            clickable: false,
            draggable: false
          })
          polygon.setOptions(polygonOptions)
          polygon.setMap(this.props.map.googleMaps)
          this.createdMapObjects.push(polygon)
          if (ringId === this.props.selectedRingId) {
            const planId = this.props.plan.activePlan.id
            const userId = this.props.user.loggedInUser.id
            var onPathChange = (path) => {
              var newPath = []
              var vertices = polygon.getPath()

              for (var i = 0; i < vertices.getLength(); i++) {
                newPath.push(vertices.getAt(i))
              }
              link.geom = newPath
              this.props.saveRingChangesToServer(ring, planId, userId)
            }

            var polygonPath = polygon.getPath()
            this.mapObjectListeners.push(
              google.maps.event.addListener(polygonPath, 'insert_at', onPathChange)
            )
            this.mapObjectListeners.push(
              google.maps.event.addListener(polygonPath, 'remove_at', onPathChange)
            )
            this.mapObjectListeners.push(
              google.maps.event.addListener(polygonPath, 'set_at', onPathChange)
            )
          }
        })

        if (ringId === this.props.selectedRingId) {
          if (ring.nodes.length > 0) {
            const coords = ring.nodes[0].data.geometry.coordinates
            var mapMarker = new google.maps.Marker({
              position: new google.maps.LatLng(coords[1], coords[0]),
              icon: {
                path: google.maps.SymbolPath.CIRCLE,
                scale: 22,
                strokeWeight: 2,
                strokeColor: '#543500'
              },
              draggable: false,
              clickable: false
            })
            mapMarker.setMap(this.props.map.googleMaps)
            this.createdMapObjects.push(mapMarker)
          }
        }

        var ringPath = new google.maps.Polyline({
          path: pathCoords,
          geodesic: true,
          strokeColor: '#543500',
          strokeOpacity: 1.0,
          strokeWeight: 1,
          clickable: false,
          draggable: false
        })
        ringPath.setMap(this.props.map.googleMaps)
        this.createdMapObjects.push(ringPath)
      }
    })
  }

  requestDeleteRing (ring) {
    swal({
      title: 'Delete Ring?',
      text: 'Are you sure you want to delete Ring: ' + ring.name + '?',
      type: 'warning',
      confirmButtonColor: '#DD6B55',
      confirmButtonText: 'Yes, delete',
      cancelButtonText: 'No',
      showCancelButton: true,
      closeOnConfirm: true
    }, (doDelete) => {
      if (doDelete) {
        const planId = this.props.plan.activePlan.id
        const userId = this.props.user.loggedInUser.id
        this.props.removeRing(ring.id, planId, userId)
      }
    })
  }

  deleteNode (ring, nodeId) {
    const planId = this.props.plan.activePlan.id
    const userId = this.props.user.loggedInUser.id
    this.props.removeNode(ring, nodeId, planId, userId)
  }

  clearRendering () {
    this.createdMapObjects.forEach(path => {
      path.setMap(null)
    })
    this.createdMapObjects = []

    this.mapObjectListeners.forEach(listener => {
      listener.remove()
    })
    this.mapObjectListeners = []
  }

  selectNewestRing () {
    var ringId = Object.keys(this.props.rings).sort().pop()
    this.props.setSelectedRingId(ringId)
  }

  componentDidMount () {
    if (!this.props.selectedRingId) {
      this.selectNewestRing()
    }
  }

  componentWillUnmount () {
    this.clearRendering()
  }
}

// --- //

RingEdit.propTypes = {
  rings: PropTypes.objectOf(PropTypes.instanceOf(Ring)),
  selectedRingId: PropTypes.number,
  plan: PropTypes.object,
  user: PropTypes.object,
  map: PropTypes.object
}

const mapStateToProps = (state) => ({
  rings: state.ringEdit.rings,
  selectedRingId: state.ringEdit.selectedRingId,
  plan: state.plan,
  user: state.user,
  map: state.map,
  status: state.plan.activePlan && state.plan.activePlan.planState,
  ringOptionsBasic: state.ringEdit.aroNetworkConstraints
})

const mapDispatchToProps = dispatch => ({
  setSelectedRingId: ringId => dispatch(ringActions.setSelectedRingId(ringId)),
  newRing: (planId, userId) => dispatch(ringActions.newRing(planId, userId)),
  removeRing: (ringId, planId, userId) => dispatch(ringActions.removeRing(ringId, planId, userId)),
  removeNode: (ring, featureId, planId, userId) => dispatch(ringActions.removeNode(ring, featureId, planId, userId)),
  saveRingChangesToServer: (ring, planId, userId) => dispatch(ringActions.saveRingChangesToServer(ring, planId, userId)),
  renameRing: (ring, name, planId, userId) => dispatch(ringActions.renameRing(ring, name, planId, userId))
})

const RingEditComponent = wrapComponentWithProvider(reduxStore, RingEdit, mapStateToProps, mapDispatchToProps)
export default RingEditComponent
