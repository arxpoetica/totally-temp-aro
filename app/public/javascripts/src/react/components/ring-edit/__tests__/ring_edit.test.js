import React from 'react'
import { shallow } from 'enzyme'
import { RingEdit } from '../ring-edit'
import RingStatusTypes from '../constants'
import Ring from '../../../common/ring'
import RingTestData from '../testData/ringTestData'

const mockSetMap = jest.fn()
const mockSetPosition = jest.fn()
const mockGetPath = jest.fn()
const mockSetOptions = jest.fn()
const mockMarkerConstructor = jest.fn(markerObj => {
  return { ...markerObj,
    setMap: mockSetMap
  }
})
const mockPolylineConstructor = jest.fn(polylineObj => {
  return { ...polylineObj,
    setMap: mockSetMap
  }
})
const mockPolygonConstructor = jest.fn(polygonObj => {
  return { ...polygonObj,
    setMap: mockSetMap,
    setOptions: mockSetOptions,
    getPath: mockGetPath
  }
})
const mockLatLngConstructor = jest.fn(latLngObj => {
  // Return an object with a mock for setPosition(), that will be called by our component
  return latLngObj
})
const mockAddListener = jest.fn()

const globalGoogle = {
  maps: {
    Marker: mockMarkerConstructor,
    Polygon: mockPolygonConstructor,
    Polyline: mockPolylineConstructor,
    LatLng: mockLatLngConstructor,
    event: {
      addListener: mockAddListener
    },
    setOptions: () => {}
  }
}

function getRings () {
  global.google = globalGoogle
  var ring48 = Ring.parseData(RingTestData.ringData48, RingTestData.ringEq48)
  var ring49 = Ring.parseData(RingTestData.ringData49, RingTestData.ringEq49)

  return {
    '48': ring48,
    '49': ring49
  }
}

//var rings = getRings()

test('With no rings', () => {
  const mockSetSelectedRingId = jest.fn()
  const component = shallow(
    <RingEdit rings={{}}
      setSelectedRingId={mockSetSelectedRingId}
    />
  )
  expect(component).toMatchSnapshot()
})

test("With rings, can't edit", () => {
  global.google = globalGoogle
  const mockSetSelectedRingId = jest.fn()
  const mockGoogleMaps = {}
  var rings = getRings()

  const component = shallow(
    <RingEdit rings={rings}
      setSelectedRingId={mockSetSelectedRingId}
      map={{googleMaps: mockGoogleMaps}}
      status={RingStatusTypes.COMPLETED}
    />
  )
  expect(component).toMatchSnapshot()
})



