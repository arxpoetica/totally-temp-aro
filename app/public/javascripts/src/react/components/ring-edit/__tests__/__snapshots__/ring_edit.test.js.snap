// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can edit 1`] = `
<div>
  <div
    className="m-2 p-2"
  >
    <button
      className="btn btn-sm btn-light"
      id="btnRingNewRing"
      onClick={[Function]}
    >
      <i
        className="fas fa-pencil-alt"
      />
       Add Ring
    </button>
  </div>
  <div
    className="m-2 p-2"
  >
    <h4>
      Rings
    </h4>
    <table
      className="table table-sm table-striped"
    >
      <tbody>
        <tr
          key="49"
        >
          <td>
            <div
              id="btnRingSelect_49"
              onClick={[Function]}
            >
              south
            </div>
          </td>
        </tr>
        <tr
          key="48"
        >
          <td
            className="ring-table-item-selected"
          >
            <div
              className="ring-table-item-title ring-table-item-title-selected clearfix"
            >
              north
              <button
                className="btn btn-sm btn-outline-danger ring-del-btn"
                data-placement="bottom"
                data-toggle="tooltip"
                id="btnRingDel_48"
                onClick={[Function]}
                title="Delete"
              >
                <i
                  className="fa ei-button-icon ng-scope fa-trash-alt"
                />
              </button>
              <input
                className="form-control form-control-sm ring-text-inp"
                id="inpRingName_48"
                onBlur={[Function]}
                onKeyDown={[Function]}
                placeholder="rename"
                type="text"
              />
            </div>
            <div
              className="ring-sub-table"
            >
              <table
                className="table table-sm table-striped"
              >
                <tbody>
                  <tr
                    className="m-2 p-2"
                    key="48_7d3495ce-59be-4900-8189-33e09dec49fb"
                  >
                    <td>
                      7d3495ce-59be-4900-8189-33e09dec49fb
                      <button
                        className="btn btn-sm btn-outline-danger ring-del-btn"
                        data-placement="bottom"
                        data-toggle="tooltip"
                        id="btnNodeDel_48-7d3495ce-59be-4900-8189-33e09dec49fb"
                        onClick={[Function]}
                        title="Delete"
                      >
                        <i
                          className="fa ei-button-icon ng-scope fa-trash-alt"
                        />
                      </button>
                    </td>
                  </tr>
                  <tr
                    className="m-2 p-2"
                    key="48_f49632eb-c747-4d7f-9797-574f7796afbd"
                  >
                    <td>
                      f49632eb-c747-4d7f-9797-574f7796afbd
                      <button
                        className="btn btn-sm btn-outline-danger ring-del-btn"
                        data-placement="bottom"
                        data-toggle="tooltip"
                        id="btnNodeDel_48-f49632eb-c747-4d7f-9797-574f7796afbd"
                        onClick={[Function]}
                        title="Delete"
                      >
                        <i
                          className="fa ei-button-icon ng-scope fa-trash-alt"
                        />
                      </button>
                    </td>
                  </tr>
                  <tr
                    className="m-2 p-2"
                    key="48_267987d6-819a-4b25-a819-9d4a7e486d74"
                  >
                    <td>
                      267987d6-819a-4b25-a819-9d4a7e486d74
                      <button
                        className="btn btn-sm btn-outline-danger ring-del-btn"
                        data-placement="bottom"
                        data-toggle="tooltip"
                        id="btnNodeDel_48-267987d6-819a-4b25-a819-9d4a7e486d74"
                        onClick={[Function]}
                        title="Delete"
                      >
                        <i
                          className="fa ei-button-icon ng-scope fa-trash-alt"
                        />
                      </button>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <ReduxForm
    destroyOnUnmount={true}
    enableReinitialize={true}
    forceUnregisterOnUnmount={false}
    form="ringOptions"
    getFormState={[Function]}
    initialValues={
      Object {
        "conduitBufferSize": Object {
          "displayName": "Conduit buffer size",
          "value": 500,
        },
        "locationBufferSize": Object {
          "displayName": "Location buffer size",
          "value": 500,
        },
        "maxConnectionDistance": Object {
          "displayName": "Connection distance",
          "value": 20,
        },
        "maxLocationEdgeDistance": Object {
          "displayName": "Max location-edge distance",
          "value": 400,
        },
        "maxWormholeDistance": Object {
          "displayName": "Wormhole distance",
          "value": 40,
        },
        "ringComplexityCount": Object {
          "displayName": "Ring complexity",
          "value": 3000000,
        },
        "snappingDistance": Object {
          "displayName": "Snapping distance",
          "value": 1,
        },
        "targetEdgeTypes": Object {
          "duct": Object {
            "displayName": "Duct",
            "value": false,
          },
          "road": Object {
            "displayName": "Road",
            "value": true,
          },
          "sewer": Object {
            "displayName": "Sewer",
            "value": true,
          },
        },
      }
    }
    keepDirtyOnReinitialize={false}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldError={[Function]}
    shouldValidate={[Function]}
    shouldWarn={[Function]}
    submitAsSideEffect={false}
    touchOnBlur={true}
    touchOnChange={false}
    updateUnregisteredFields={false}
  />
</div>
`;

exports[`Can't edit 1`] = `
<div>
  <div
    className="m-2 p-2"
  >
    <h4>
      Rings
    </h4>
    <table
      className="table table-sm table-striped"
    >
      <tbody>
        <tr
          key="49"
        >
          <td>
            <div
              id="btnRingSelect_49"
              onClick={[Function]}
            >
              south
            </div>
          </td>
        </tr>
        <tr
          key="48"
        >
          <td
            className="ring-table-item-selected"
          >
            <div
              className="ring-table-item-title ring-table-item-title-selected clearfix"
            >
              north
            </div>
            <div
              className="ring-sub-table"
            >
              <table
                className="table table-sm table-striped"
              >
                <tbody>
                  <tr
                    className="m-2 p-2"
                    key="48_7d3495ce-59be-4900-8189-33e09dec49fb"
                  >
                    <td>
                      7d3495ce-59be-4900-8189-33e09dec49fb
                    </td>
                  </tr>
                  <tr
                    className="m-2 p-2"
                    key="48_f49632eb-c747-4d7f-9797-574f7796afbd"
                  >
                    <td>
                      f49632eb-c747-4d7f-9797-574f7796afbd
                    </td>
                  </tr>
                  <tr
                    className="m-2 p-2"
                    key="48_267987d6-819a-4b25-a819-9d4a7e486d74"
                  >
                    <td>
                      267987d6-819a-4b25-a819-9d4a7e486d74
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <ReduxForm
    destroyOnUnmount={true}
    enableReinitialize={true}
    forceUnregisterOnUnmount={false}
    form="ringOptions"
    getFormState={[Function]}
    initialValues={
      Object {
        "conduitBufferSize": Object {
          "displayName": "Conduit buffer size",
          "value": 500,
        },
        "locationBufferSize": Object {
          "displayName": "Location buffer size",
          "value": 500,
        },
        "maxConnectionDistance": Object {
          "displayName": "Connection distance",
          "value": 20,
        },
        "maxLocationEdgeDistance": Object {
          "displayName": "Max location-edge distance",
          "value": 400,
        },
        "maxWormholeDistance": Object {
          "displayName": "Wormhole distance",
          "value": 40,
        },
        "ringComplexityCount": Object {
          "displayName": "Ring complexity",
          "value": 3000000,
        },
        "snappingDistance": Object {
          "displayName": "Snapping distance",
          "value": 1,
        },
        "targetEdgeTypes": Object {
          "duct": Object {
            "displayName": "Duct",
            "value": false,
          },
          "road": Object {
            "displayName": "Road",
            "value": true,
          },
          "sewer": Object {
            "displayName": "Sewer",
            "value": true,
          },
        },
      }
    }
    keepDirtyOnReinitialize={false}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldError={[Function]}
    shouldValidate={[Function]}
    shouldWarn={[Function]}
    submitAsSideEffect={false}
    touchOnBlur={true}
    touchOnChange={false}
    updateUnregisteredFields={false}
  />
</div>
`;

exports[`Delete ring 1`] = `
<div>
  <div
    className="m-2 p-2"
  >
    <button
      className="btn btn-sm btn-light"
      id="btnRingNewRing"
      onClick={[Function]}
    >
      <i
        className="fas fa-pencil-alt"
      />
       Add Ring
    </button>
  </div>
  <div
    className="m-2 p-2"
  >
    <h4>
      Rings
    </h4>
    <table
      className="table table-sm table-striped"
    >
      <tbody>
        <tr
          key="49"
        >
          <td>
            <div
              id="btnRingSelect_49"
              onClick={[Function]}
            >
              south
            </div>
          </td>
        </tr>
        <tr
          key="48"
        >
          <td
            className="ring-table-item-selected"
          >
            <div
              className="ring-table-item-title ring-table-item-title-selected clearfix"
            >
              north
              <button
                className="btn btn-sm btn-outline-danger ring-del-btn"
                data-placement="bottom"
                data-toggle="tooltip"
                id="btnRingDel_48"
                onClick={[Function]}
                title="Delete"
              >
                <i
                  className="fa ei-button-icon ng-scope fa-trash-alt"
                />
              </button>
              <input
                className="form-control form-control-sm ring-text-inp"
                id="inpRingName_48"
                onBlur={[Function]}
                onKeyDown={[Function]}
                placeholder="rename"
                type="text"
              />
            </div>
            <div
              className="ring-sub-table"
            >
              <table
                className="table table-sm table-striped"
              >
                <tbody>
                  <tr
                    className="m-2 p-2"
                    key="48_7d3495ce-59be-4900-8189-33e09dec49fb"
                  >
                    <td>
                      7d3495ce-59be-4900-8189-33e09dec49fb
                      <button
                        className="btn btn-sm btn-outline-danger ring-del-btn"
                        data-placement="bottom"
                        data-toggle="tooltip"
                        id="btnNodeDel_48-7d3495ce-59be-4900-8189-33e09dec49fb"
                        onClick={[Function]}
                        title="Delete"
                      >
                        <i
                          className="fa ei-button-icon ng-scope fa-trash-alt"
                        />
                      </button>
                    </td>
                  </tr>
                  <tr
                    className="m-2 p-2"
                    key="48_f49632eb-c747-4d7f-9797-574f7796afbd"
                  >
                    <td>
                      f49632eb-c747-4d7f-9797-574f7796afbd
                      <button
                        className="btn btn-sm btn-outline-danger ring-del-btn"
                        data-placement="bottom"
                        data-toggle="tooltip"
                        id="btnNodeDel_48-f49632eb-c747-4d7f-9797-574f7796afbd"
                        onClick={[Function]}
                        title="Delete"
                      >
                        <i
                          className="fa ei-button-icon ng-scope fa-trash-alt"
                        />
                      </button>
                    </td>
                  </tr>
                  <tr
                    className="m-2 p-2"
                    key="48_267987d6-819a-4b25-a819-9d4a7e486d74"
                  >
                    <td>
                      267987d6-819a-4b25-a819-9d4a7e486d74
                      <button
                        className="btn btn-sm btn-outline-danger ring-del-btn"
                        data-placement="bottom"
                        data-toggle="tooltip"
                        id="btnNodeDel_48-267987d6-819a-4b25-a819-9d4a7e486d74"
                        onClick={[Function]}
                        title="Delete"
                      >
                        <i
                          className="fa ei-button-icon ng-scope fa-trash-alt"
                        />
                      </button>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <ReduxForm
    destroyOnUnmount={true}
    enableReinitialize={true}
    forceUnregisterOnUnmount={false}
    form="ringOptions"
    getFormState={[Function]}
    initialValues={
      Object {
        "conduitBufferSize": Object {
          "displayName": "Conduit buffer size",
          "value": 500,
        },
        "locationBufferSize": Object {
          "displayName": "Location buffer size",
          "value": 500,
        },
        "maxConnectionDistance": Object {
          "displayName": "Connection distance",
          "value": 20,
        },
        "maxLocationEdgeDistance": Object {
          "displayName": "Max location-edge distance",
          "value": 400,
        },
        "maxWormholeDistance": Object {
          "displayName": "Wormhole distance",
          "value": 40,
        },
        "ringComplexityCount": Object {
          "displayName": "Ring complexity",
          "value": 3000000,
        },
        "snappingDistance": Object {
          "displayName": "Snapping distance",
          "value": 1,
        },
        "targetEdgeTypes": Object {
          "duct": Object {
            "displayName": "Duct",
            "value": false,
          },
          "road": Object {
            "displayName": "Road",
            "value": true,
          },
          "sewer": Object {
            "displayName": "Sewer",
            "value": true,
          },
        },
      }
    }
    keepDirtyOnReinitialize={false}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldError={[Function]}
    shouldValidate={[Function]}
    shouldWarn={[Function]}
    submitAsSideEffect={false}
    touchOnBlur={true}
    touchOnChange={false}
    updateUnregisteredFields={false}
  />
</div>
`;

exports[`With no rings 1`] = `
<div>
  <div
    className="m-2 p-2"
  >
    <h4>
      Rings
    </h4>
    <table
      className="table table-sm table-striped"
    >
      <tbody />
    </table>
  </div>
  <ReduxForm
    destroyOnUnmount={true}
    enableReinitialize={true}
    forceUnregisterOnUnmount={false}
    form="ringOptions"
    getFormState={[Function]}
    initialValues={
      Object {
        "conduitBufferSize": Object {
          "displayName": "Conduit buffer size",
          "value": 500,
        },
        "locationBufferSize": Object {
          "displayName": "Location buffer size",
          "value": 500,
        },
        "maxConnectionDistance": Object {
          "displayName": "Connection distance",
          "value": 20,
        },
        "maxLocationEdgeDistance": Object {
          "displayName": "Max location-edge distance",
          "value": 400,
        },
        "maxWormholeDistance": Object {
          "displayName": "Wormhole distance",
          "value": 40,
        },
        "ringComplexityCount": Object {
          "displayName": "Ring complexity",
          "value": 3000000,
        },
        "snappingDistance": Object {
          "displayName": "Snapping distance",
          "value": 1,
        },
        "targetEdgeTypes": Object {
          "duct": Object {
            "displayName": "Duct",
            "value": false,
          },
          "road": Object {
            "displayName": "Road",
            "value": true,
          },
          "sewer": Object {
            "displayName": "Sewer",
            "value": true,
          },
        },
      }
    }
    keepDirtyOnReinitialize={false}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldError={[Function]}
    shouldValidate={[Function]}
    shouldWarn={[Function]}
    submitAsSideEffect={false}
    touchOnBlur={true}
    touchOnChange={false}
    updateUnregisteredFields={false}
  />
</div>
`;
